### PR TITLE
Add Bimi feature

### PR DESCRIPTION
**Pull request by NicolasBourdin88**
We need to be careful about when they will add format PNG to the route, normally it will not affect our route because they (IOS) will just have to add "?format=png" at the end of the call.

Depends on https://github.com/Infomaniak/android-core/pull/179

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/data/models/Bimi.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Parcelize
+@Serializable
+class Bimi() : EmbeddedRealmObject, Parcelable {
+    @SerialName("svg_content")
+    var svgContentUrl: String? = null
+    @SerialName("is_certified")
+    var isCertified: Boolean = false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/models/Bimi.kt:34]:
```suggestion

    //region Remote data
    @SerialName("svg_content")
    var svgContentUrl: String? = null
    @SerialName("is_certified")
    var isCertified: Boolean = false
    //endregion
```

File: app/src/main/java/com/infomaniak/mail/ui/main/AvatarNameEmailView.kt [hunk 2]
```diff
@@ -86,12 +87,17 @@ class AvatarNameEmailView @JvmOverloads constructor(
         setNameAndEmail(attendee)
     }
 
-    private fun ViewAvatarNameEmailBinding.setNameAndEmail(correspondent: Correspondent) {
+    private fun ViewAvatarNameEmailBinding.setNameAndEmail(
+        correspondent: Correspondent,
+        isCorrespondentCertified: Boolean = false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/AvatarNameEmailView.kt:92]:
```suggestion
        isCorrespondentCertified: Boolean = false,
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListAdapter.kt [hunk 3]
```diff
@@ -400,7 +401,12 @@ class ThreadListAdapter @Inject constructor(
     }
 
     private fun CardviewThreadItemBinding.displayAvatar(thread: Thread) {
-        expeditorAvatar.loadAvatar(thread.computeAvatarRecipient())
+        val (recipient, bimi) = thread.computeAvatarRecipient()
+        if (bimi != null && bimi.isCertified) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListAdapter.kt:405]:
```suggestion
        if (bimi?.isCertified == true) {
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 4]
```diff
@@ -121,7 +122,7 @@ class ThreadAdapter(
                 shouldLoadDistantResources,
                 threadAdapterCallbacks?.onContactClicked,
                 threadAdapterCallbacks?.onAttachmentClicked,
-                threadAdapterCallbacks?.onAttachmentOptionsClicked,
+                threadAdapterCallbacks?.onAttachmentOptionsClicked
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:125]: Revert this change.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt [hunk 5]
```diff
@@ -23,17 +23,18 @@ import androidx.recyclerview.widget.RecyclerView.Adapter
...
 import com.infomaniak.mail.MatomoMail.trackMessageEvent
+import com.infomaniak.mail.data.models.Bimi
 import com.infomaniak.mail.data.models.correspondent.Recipient
 import com.infomaniak.mail.databinding.ItemDetailedContactBinding
 import com.infomaniak.mail.ui.main.thread.DetailedRecipientAdapter.DetailedRecipientViewHolder
 import com.infomaniak.mail.utils.UiUtils.fillInUserNameAndEmail
 
 class DetailedRecipientAdapter(
-    private val onContactClicked: ((contact: Recipient) -> Unit)?,
+    private val onContactClicked: ((contact: Recipient, bimi: Bimi?) -> Unit)?
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt:33]: Revert this comma removal.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt [hunk 6]
```diff
@@ -23,17 +23,18 @@ import androidx.recyclerview.widget.RecyclerView.Adapter
...
 ) : Adapter<DetailedRecipientViewHolder>() {
 
     private var recipients = emptyList<Recipient>()
-
+    private var bimi: Bimi? = null
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt:37]: Revert this jump line removal.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 7]
```diff
@@ -224,10 +224,13 @@ class ThreadFragment : Fragment() {
...
             threadAdapterCallbacks = ThreadAdapterCallbacks(
-                onContactClicked = {
+                onContactClicked =  { recipient, bimi ->
                     safeNavigate(
                         resId = R.id.detailedContactBottomSheetDialog,
-                        args = DetailedContactBottomSheetDialogArgs(it).toBundle(),
+                        args = DetailedContactBottomSheetDialogArgs(
+                            recipient = recipient,
+                            bimi
+                        ).toBundle(),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:233]: You can one-line it back if you remove `recipient` parameter name.

File: app/src/main/java/com/infomaniak/mail/views/AvatarView.kt [hunk 8]
```diff
@@ -55,14 +58,20 @@ class AvatarView @JvmOverloads constructor(
     private val binding by lazy { ViewAvatarBinding.inflate(LayoutInflater.from(context), this, true) }
 
     private var currentCorrespondent: Correspondent? = null
+    private var isBimiShow: Boolean = false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/AvatarView.kt:61]: Shouldn't it be named `isBimiShown` ?

File: app/src/main/java/com/infomaniak/mail/views/AvatarView.kt [hunk 9]
```diff
@@ -140,6 +149,21 @@ class AvatarView @JvmOverloads constructor(
         binding.avatarImage.load(R.drawable.ic_unknown_user_avatar)
     }
 
+    fun loadBimiAvatar(urlBimi: String, correspondent: Correspondent?) = with(binding.avatarImage) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/AvatarView.kt:152]: Rename `urlBimi` into `bimiUrl`.

File: app/src/main/java/com/infomaniak/mail/views/AvatarView.kt [hunk 10]
```diff
@@ -140,6 +149,21 @@ class AvatarView @JvmOverloads constructor(
...
+        contentDescription = correspondent?.email.orEmpty()
+        isBimiShow = urlBimi.isNotEmpty()
+        loadAvatar(
+            backgroundColor = context.getBackgroundColorBasedOnId(
+                correspondent?.email.orEmpty().hashCode(),
+                R.array.AvatarColors
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/views/AvatarView.kt:158]:
```suggestion
                R.array.AvatarColors,
```

File: app/src/main/res/drawable/ic_certified.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~ Infomaniak Mail - Android
+  ~ Copyright (C) 2023-2024 Infomaniak Network SA
```
Comment (KevinBoulongne) [inline app/src/main/res/drawable/ic_certified.xml:3]:
```suggestion
  ~ Copyright (C) 2024 Infomaniak Network SA
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 12]
```diff
@@ -712,9 +717,9 @@ class ThreadAdapter(
     private class MessageViewHolder(
         override val binding: ItemMessageBinding,
         private val shouldLoadDistantResources: Boolean,
-        onContactClicked: ((contact: Recipient) -> Unit)?,
+        onContactClicked: ((contact: Recipient, bimi: Bimi?) -> Unit)?,
         onAttachmentClicked: ((attachment: Attachment) -> Unit)?,
-        onAttachmentOptionsClicked: ((attachment: Attachment) -> Unit)?,
+        onAttachmentOptionsClicked: ((attachment: Attachment) -> Unit)?
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:722]:
```suggestion
        onAttachmentOptionsClicked: ((attachment: Attachment) -> Unit)?,
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt [hunk 13]
```diff
@@ -18,10 +18,7 @@
 package com.infomaniak.mail.data.cache
 
 import android.content.Context
-import com.infomaniak.mail.data.models.AppSettings
-import com.infomaniak.mail.data.models.Attachment
-import com.infomaniak.mail.data.models.Folder
-import com.infomaniak.mail.data.models.Quotas
+import com.infomaniak.mail.data.models.*
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt:21]: We avoid wildcards import.

File: app/src/main/java/com/infomaniak/mail/views/AvatarView.kt [hunk 14]
```diff
@@ -167,4 +191,12 @@ class AvatarView @JvmOverloads constructor(
             )
         }
     }
+
+    fun loadAvatar(correspondent: Correspondent?, bimi: Bimi?) {
+        if (bimi == null || !bimi.isCertified) {
+            loadAvatar(correspondent)
+        } else {
+            loadBimiAvatar(ApiRoutes.bimi(bimi.svgContentUrl.orEmpty()), correspondent)
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/views/AvatarView.kt:199]: If `bimi.svgContentUrl` is empty, we should call `loadAvatar(correspondent)`right ?

**Review by tevincent — CHANGES_REQUESTED** [review]

**Review by tevincent — APPROVED** [review]

**Comment by tevincent** [comment]
@KevinBoulongne can you check again if everything is good for you?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt [hunk 15]
```diff
@@ -45,14 +48,15 @@ class DetailedRecipientAdapter(
...
+            onContactClicked?.invoke(recipient, bimi)
         }
     }
 
     override fun getItemCount(): Int = recipients.count()
 
-    fun updateList(newList: List<Recipient>) {
+    fun updateList(newList: List<Recipient>, newBimi: Bimi? = null) {
         recipients = newList
+        this.bimi = newBimi
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/DetailedRecipientAdapter.kt:59]:
```suggestion
        bimi = newBimi
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**